model.py

```diff
@@ -388,3 +389,55 @@ def serialize(t):
         # write to binary file
         f.close()
         print(f"wrote {filepath}")
+
+
+class LoraLinear(nn.Module):
+    """
+    LoRA: Low-Rank Adaptation of Large Language Models - https://arxiv.org/abs/2106.09685
+
+    Heavily inspired by minLoRA - https://github.com/cccntu/minLoRA
+
+    Leverages the parametrizations feature from pytorch. This allows us to add the LoRA
+    matrices to the weights during the forward pass rather than computing the modified
+    forward pass explicitly, i.e., we compute (W + BA)x rather than Wx + BAx.
+    """
+
+    def __init__(self, fan_in, fan_out, rank=4, dropout_p=0.0, alpha=1.0):
+        super().__init__()
+        self.fan_in = fan_in
+        self.fan_out = fan_out
+        self.rank = rank
+        self.dropout_p = dropout_p
+
+        self.lora_a = nn.Parameter(torch.zeros(rank, fan_in))
+        self.lora_b = nn.Parameter(torch.zeros(fan_out, rank))
+
+        nn.init.kaiming_uniform_(self.lora_a, a=math.sqrt(5))
+
+        self.scaling = alpha / rank
+        self.dropout = nn.Dropout(dropout_p)
+
+    def forward(self, weight):
+        return weight + torch.matmul(self.lora_b, self.dropout(self.lora_a)) * self.scaling
+
+
+def apply_lora(model: nn.Module, layer_types=[nn.Linear], rank=8, dropout=0.0, alpha=1.0):
+    def _apply_lora(module):
+        if type(module) in layer_types and hasattr(module, 'weight'):
+            fan_out, fan_in = module.weight.shape
+            parametrize.register_parametrization(module, 'weight', LoraLinear(fan_in, fan_out, rank, dropout, alpha))
```

Comment:
would this fail if LayerTypes is nn.Embedding? It shouldn't get replaced with LoraLinear right?